codalab/worker/image_manager.py

```diff
@@ -0,0 +1,225 @@
+        pass
+
+    def get(self, image_spec: str) -> ImageAvailabilityState:
+        """
+        Always request the newest image from the cloud if it's not in downloading thread and return the current
+        downloading status(READY, FAILED, or DOWNLOADING).
+        When the requested image in the following states:
+        1. If it's not available on the platform, we download the image and return DOWNLOADING status.
+        2. If another thread is actively downloading it, we return DOWNLOADING status.
+        3. If another thread was downloading it but not active by the time the request was sent, we return the following status:
+            * READY if the image was downloaded successfully.
+            * FAILED if the image wasn't able to be downloaded due to any reason.
+        Args:
+            image_spec: the image that the requester needs.
+                The caller will need to determine the type of image they need before calling this function.
+                It is usually safe to prefix the image with the type of image.
+                For example, the docker image go would be docker://go
+
+        Returns:
+            ImageAvailabilityState of the image requested.
+        """
+        try:
+            if image_spec in self._downloading:
+                with self._downloading[image_spec]['lock']:
+                    if self._downloading[image_spec].is_alive():
+                        return ImageAvailabilityState(
+                            digest=None,
+                            stage=DependencyStage.DOWNLOADING,
+                            message=self._downloading[image_spec]['status'],
+                        )
+                    else:
+                        if self._downloading[image_spec]['success']:
+                            status = self._image_availability_state(
+                                image_spec,
+                                success_message='Image ready',
+                                failure_message='Image {} was downloaded successfully, '
+                                'but it cannot be found locally due to unhandled error %s'.format(
+                                    image_spec
+                                ),
+                            )
+                        else:
+                            status = self._image_availability_state(
+                                image_spec,
+                                success_message='Image {} can not be downloaded from the cloud '
+                                'but it is found locally'.format(image_spec),
+                                failure_message=self._downloading[image_spec]['message'] + ": %s",
+                            )
+                        self._downloading.remove(image_spec)
+                        return status
+            else:
+                if self._max_image_size:
+                    try:
+                        image_size_bytes = self._image_size_without_pulling(image_spec)
```

Comment:
NotImplementedError nicer than returning None

```suggestion
                        try: 
                            image_size_bytes = self._image_size_without_pulling(image_spec)
                        except NotImplementedError:
                            ...
```